### ISSUE TITLE
Port several interesting&great ideas from sjpeg?

### ISSUE THREAD

**Issue by magicgoose** [mozilla/mozjpeg#274]
https://github.com/webmproject/sjpeg

• It can automatically evaluate the distortion which would be added by chroma subsampling and enable it (only)if the distortion is tolerable, this is called 'Riskiness' score
• If I understood it right, it can find optimal quantization matrices for each image, instead of using a fixed quantization table (scaled with quality). Not sure if mozjpeg already does this?

**Comment by sp-1234** [comment]
The first point as implemented there right now doesn't produce useful decisions, but the idea is good

**Comment by pornel** [comment]
Yeah, these are great experiments to learn from.

Currently cjpeg hardcodes subsampling based on quality, so image analysis could certainly improve things. I agree with @sp-1234 that the current implementation in sjpeg is not ideal.

Per-image qtables are most promising here. I haven't looked at the implementation yet.

**Comment by pornel** [comment]
Is it safe to automatically change subsampling in the libjpeg API? Can you help me investigate what could go wrong?

I'm sure it can't be done when raw data access is enabled, since it changes layout of image data. But is it safe for RGB input?

**Comment by sp-1234** [comment]
> Per-image qtables are most promising here

yeah. I compared sjpeg to mozjpeg on 1 image (both without chroma subsampling and q=90);  
sjpeg (after further crunching with mozjpeg-jpegtran) achieves 10% smaller file, while having a little bit better quality (according to dssim comparison with original image — 0.003707 vs 0.003921).
I tried to also match dssim closer, but the quality difference for ±1 is too big, and apparently sjpeg doesn't understand fractional quality values.
